batch/batch/cloud/gcp/worker/worker_api.py

```diff
@@ -128,3 +138,111 @@ async def close(self):
 
     def __str__(self):
         return f'project={self.project} zone={self.zone}'
+
+
+ContainerCredentials = Dict[str, aiogoogle.GoogleServiceAccountCredentials]
+
+
+class GoogleHailMetadataServer(HailMetadataServer[GCPUserCredentials, ContainerCredentials]):
+    def __init__(self, project: str, http_session: httpx.ClientSession):
+        super().__init__()
+        self._project = project
+        self._metadata_server_client = aiogoogle.GoogleMetadataServerClient(http_session)
+
+    def _create_container_credentials(self, default_credentials: GCPUserCredentials):
+        default_sa_creds = aiogoogle.GoogleServiceAccountCredentials(orjson.loads(default_credentials.key))
+        credentials = {}
+        for name in ('default', default_sa_creds.email):
+            credentials[name] = default_sa_creds
+        return credentials
+
+    async def _close_container_credentials(self, container_credentials: ContainerCredentials):
+        for credential in container_credentials.values():
+            await credential.close()
+
+    def _container_credentials(self, request: web.Request) -> ContainerCredentials:
+        return request['credentials']
+
+    def _user_credentials(self, request: web.Request) -> aiogoogle.GoogleServiceAccountCredentials:
+        email = request.match_info.get('gsa') or 'default'
+        return self._container_credentials(request)[email]
+
+    async def root(self, _):
+        return web.Response(text='computeMetadata/\n')
+
+    async def project_id(self, _):
+        return web.Response(text=self._project)
+
+    async def numeric_project_id(self, _):
+        return web.Response(text=await self._metadata_server_client.numeric_project_id())
+
+    async def service_accounts(self, request: web.Request):
+        accounts = '\n'.join(self._container_credentials(request).keys())
+        return web.Response(text=f'{accounts}\n')
+
+    async def user_service_account(self, request: web.Request):
+        gsa_email = self._user_credentials(request).email
+        recursive = request.query.get('recursive')
+        if recursive == 'true':
+            return web.json_response(
+                {
+                    'aliases': ['default'],
+                    'email': gsa_email,
+                    'scopes': ['https://www.googleapis.com/auth/cloud-platform'],
+                },
+            )
+        return web.Response(text='aliases\nemail\nidentity\nscopes\ntoken\n')
```

Comment:
Hmm. This feels a bit out of sync. You implement:
- `service-accounts/{name}/`
- `service-accounts/{name}/email`
- `service-accounts/{name}/token`

But this suggests that there is also aliases, identity, and scopes.

Also do we need to also serve `service-accounts/{name}`?

Should recursive include identity and token?

Reply:
> But this suggests that there is also aliases, identity, and scopes.

I've included only the endpoints that `gcloud` seems to use for basic operations. We could expand functionality as necessary but this isn't intended to be a complete implementation, just enough for providing an authentication mechanism for hail tools and `gcloud`.

> Also do we need to also serve service-accounts/{name}?

This would just redirect to `service-accounts/{name}/` and didn't seem to be used by `gcloud` so I didn't bother implementing it.

> Should recursive include identity and token?

From manual inspection, it looks like `recursive` excludes those endpoints, possibly because they deliver secrets or return resources that are not static metadata.